modules/imgcodecs/src/grfmt_pfm.cpp

```diff
@@ -0,0 +1,314 @@
+  for (size_t i = 0; i < buffer_size; ++i) {
+    char c = strm.getByte();
+    if (std::isspace(c)) {
+      break;
+    }
+    buffer[i] = c;
+  }
+  const std::string str(buffer.begin(), buffer.end());
+  return atoT<T>(str);
+}
+
+template<typename T> void write_anything(cv::WLByteStream& strm, const T& t)
+{
+  const std::string str = std::to_string(t);
+  strm.putBytes(str.c_str(), str.size());
+}
+
+}
+
+namespace cv {
+
+PFMDecoder::~PFMDecoder()
+{
+}
+
+PFMDecoder::PFMDecoder()
+{
+  m_strm.close();
+}
+
+bool PFMDecoder::readHeader()
+{
+  if (m_buf.empty()) {
+    if (!m_strm.open(m_filename)) {
+      return false;
+    }
+  } else {
+    if (!m_strm.open(m_buf)) {
+      return false;
+    }
+  }
+
+  if (m_strm.getByte() != 'P') {
+    CV_Error(Error::StsError, "Unexpected file type (expected P)");
+    return false;
+  }
+
+  switch (m_strm.getByte()) {
+  case 'f':
+    m_channels = 1;
+    break;
+  case 'F':
+    m_channels = 3;
+    break;
+  default:
+    CV_Error(Error::StsError, "Unexpected file type (expected `f` or `F`)");
+    return false;
+  }
+
+  if ('\n' != m_strm.getByte()) {
+    CV_Error(Error::StsError, "Unexpected header format (expected line break)");
+    return false;
+  }
+
+
+  m_width = read_number<int>(m_strm);
+  m_height = read_number<int>(m_strm);
+  m_scale_factor = read_number<float>(m_strm);
+  m_swap_byte_order = is_byte_order_swapped(m_scale_factor);
+
+  return true;
+}
+
+bool PFMDecoder::readData(Mat& mat)
+{
+  if (!m_strm.isOpened()) {
+    CV_Error(Error::StsError, "Unexpected status in data stream");
+    return false;
+  }
+
+  if (m_channels == 3) {
+    mat = Mat3f(m_height, m_width);
+  } else if (m_channels == 1) {
+    mat = Mat1f(m_height, m_width);
+  } else {
+    CV_Error(Error::StsError, "Expected 1 or 3 channels.");
+    return false;
+  }
+
+  CV_Assert(abs(m_scale_factor) == 1.0);
+
+  for (int y = m_height - 1; y >= 0; --y) {
+    m_strm.getBytes(mat.ptr(y), m_width * mat.elemSize());
+    if (is_byte_order_swapped(m_scale_factor)) {
+      for (size_t i = 0; i < m_width * m_channels; ++i) {
+        swap_endianess(mat.ptr<float>(y)[i]);
+      }
+    }
+  }
+
+  if (mat.channels() == 3) {
+    cv::cvtColor(mat, mat, cv::COLOR_BGR2RGB);
+  }
+
+  mat /= abs(m_scale_factor);
+
+
+  return true;
+}
+
+size_t PFMDecoder::signatureLength() const
+{
+    return 2;
+}
+
+bool PFMDecoder::checkSignature( const String& signature ) const
+{
+    return signature.size() >= 2 && signature[0] == 'P'
+                                 && ( signature[1] == 'f' || signature[1] == 'F' );
+}
+
+void PFMDecoder::close()
+{
+  // noop
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////
+
+PFMEncoder::PFMEncoder()
+{
+  m_description = "Portable image format - float (*.pfm)";
+}
+
+PFMEncoder::~PFMEncoder()
+{
+}
+
+bool PFMEncoder::isFormatSupported(int depth) const
+{
+    return depth == CV_32FC1 || depth == CV_32FC3;
```

Comment:
Depth doesn't contain information about number of channels. So `CV_32FC3` is a bad value here.
Valid depth values is [here](https://github.com/opencv/opencv/blob/3.4.2/modules/core/include/opencv2/core/hal/interface.h#L71-L77)  (C1 values are mapped to depth directly, i.e `CV_32FC1 == CV_32F`)

Reply:
See this comment: https://github.com/opencv/opencv/pull/12192#issuecomment-412051004

You can leave this "as is", but you should not pass 8U images in tests.